### PR TITLE
Gui: Add option to show workbench selector in menu bar

### PR DESCRIPTION
**Pull request by PaddleStroke**
**Problem definition:** The workbench selector is not displayed prominently, making it difficult for new users to find and use.

**Proposed solution:** Allow the workbench selector to be displayed in the menu bar. This requires the addition of a preference to control location of the workbench so that the current behaviour can be retained or the new, easier to use behaviour applied for new users.

**PR description:** 
This enables the user to select where he wants to have the workbench selector:
- in toolbar (current behaviour)
- In menubar on the left
- in menubar on the right.

This settings requires a restart of the app (same as 'tree view mode' for instance).

Discussion here : https://forum.freecadweb.org/viewtopic.php?f=8&t=72431&p=632779#p632779

Note: The previous PR for this feature has been closed. The branch had issue so I deleted and restarted.

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [x]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [x]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [x]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [x]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the issue ID number from our [Issues database](https://github.com/FreeCAD/FreeCAD/issues) in case a particular commit solves or is related to an existing issue. Ex: `Draft: fix typos - fixes #4805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

---

### CODE REVIEW

**Comment by PaddleStroke** [comment]
@0penBrain I opened a new PR for this feature because the other branch was causing problem for some reason and I deleted it. 

I simplified a lot this PR following your comments @0penBrain. I dropped the idea of having the setting being taken into account on-the-fly, which was the root of many of your comments due to dubious hacks.
So now the user has to restart the app for the setting to be working. Which is the case for 'Tree mode view' setting for instance. I wrote it similarly in the tooltip.

@chennes is that kind of PR something you could look at? It's now much simpler and should not be controversial as it doesn't change anything by default.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7679">pipeline status</a> for feature branch [PR_7679](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7679). Pipeline [682581827](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/682581827) was triggered at [13568c8](https://github.com/FreeCAD/FreeCAD/pull/7679/commits/13568c8bb37210c453f488e6534b1f73a017800c). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7679">pipeline status</a> for feature branch [PR_7679](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7679). Pipeline [682606991](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/682606991) was triggered at [3c0fdbd](https://github.com/FreeCAD/FreeCAD/pull/7679/commits/3c0fdbd12a67f26cda42bd7f48b2b62532ab3efd). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Gui/DlgGeneral.ui [hunk 1]
```diff
@@ -505,6 +505,43 @@ this according to your screen size or personal taste</string>
...
+           <item>
+            <widget class="QComboBox" name="WorkbenchSelectorPosition">
+             <property name="toolTip">
+              <string>Customize where the workbench selector appears (restart required).
+
+'Toolbar': In the toolbars, as a movable toolbar.
+'Left Corner': In the menu bar, on the left corner.
+'Right Corner': In the menu bar, on the right corne.</string>
+             </property>
+            </widget>
```
Comment (0penBrain) [inline src/Gui/DlgGeneral.ui:541]: Why not use a `PrefComboBox` ? Why not populate the items here ?

Typo in the tooltip : `right corner.`

Comment (PaddleStroke) [reply src/Gui/DlgGeneral.ui:541]: I reused the syntax from the above, treeMode combobox, such that the code is more homogeneous.

File: src/Gui/Action.cpp [hunk 2]
```diff
@@ -539,6 +540,24 @@ void WorkbenchGroup::addTo(QWidget *w)
...
+        auto menuBar = qobject_cast<QMenuBar*>(w);
+        ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/MainWindow");
+        
+        QComboBox* box = new WorkbenchComboBox(this, w);
+        box->setIconSize(QSize(16, 16));
+        box->setToolTip(_action->toolTip());
+        box->setStatusTip(_action->statusTip());
+        box->setWhatsThis(_action->whatsThis());
+        box->addActions(_group->actions());
+        connect(_group, SIGNAL(triggered(QAction*)), box, SLOT(onActivated(QAction*)));
```
[the remark this answers is not in the document]
Comment (PaddleStroke) [reply src/Gui/Action.cpp:553]: Ok done.

File: src/Gui/DlgGeneral.ui (hunk 1 above)
Comment (PaddleStroke) [reply src/Gui/DlgGeneral.ui:541]: If it's not the best way to do things, maybe a next PR can change that for all comboboxes.

**Comment by PaddleStroke** [comment]
Just forced push with @0penBrain corrections.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7679">pipeline status</a> for feature branch [PR_7679](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7679). Pipeline [683888277](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/683888277) was triggered at [b72c404](https://github.com/FreeCAD/FreeCAD/pull/7679/commits/b72c404ac56f4f13669f16ca3a974546153a4dfb). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by PaddleStroke** [comment]
@wwmayer I think this kind of Gui setting PR is in your realm. Do you want to have a look at it ?

**Comment by sliptonic** [comment]
In keeping with our attempt to move to a transparent contribution process, could you comment on what problem this PR is solving?

**Comment by PaddleStroke** [comment]
UI improvement do not necessarily resolve a problem per say.
Though the initial idea was to make the workbench selector more prominent.
To make it more obvious for new users.

As in current situation the workbench selector is in the middle of the
toolbars.
Myself I remember on my first launch of freecad, I had trouble
understanding how FC worked and to change workbench.

So my goal was to move the workbench selector to the menu bar.

However it will be impossible to have this accepted by the veterans of the
forum. Especially since some Linux users are using something called global
menu which modify the menu bar.

So in the end I just made it an option.

Long story short, the advantages :
- Save space
- More prominent.

**Comment by silopolis** [comment]
Le dim. 6 nov. 2022 à 18:58, PaddleStroke ***@***.***> a
écrit :

> UI improvement do not necessarily resolve a problem per say.
> Though the initial idea was to make the workbench selector more prominent.
> To make it more obvious for new users.
>


I for one also find this perfectly sound with the workbench paradigm in
freecad.

My favorite choice would be to have an button before the file menu with the
icon of the active wb, opening a menu to access the list and change wb...
Or that could be the File menu as can be seen in other softwares in the
field, with the wb menu next to it then the generic ones and finally the
workbench(es) specific menus, before the usual windows and help menus.


As in current situation the workbench selector is in the middle of the
> toolbars.



Which I find annoying.


Second best option IMHO is the tabbed paradigm

My 2 user cts

**Comment by WandererFan** [comment]
Problem definition: The workbench selector is not displayed prominently, making it difficult for new users to find and use.

Proposed solution: Allow the workbench selector to be displayed in the menu bar.  This requires the addition of a preference to control location of the workbench so that the current behaviour can be retained or the new, easier to use behaviour applied for new users.

PR Description:  This PR implements the proposed solution for problem XXXX.

This way the merger only has to see that there exists a problem and that the PR solves it instead of looking through 6 pages of forum posts to learn the context.

**Comment by WandererFan** [comment]
Could not compile this.  Version with these changes:
Missing #include <QMenuBar> in Action.cpp
rebase on master
Missing #include <FCGlobal.h> in Action.h
Missing #include <QComboBox> in MenuManager.cpp
is here:  https://github.com/WandererFan/FreeCAD/tree/WB_cornerwidget

Warning I didn't look into:
[ 61%] Building CXX object src/Gui/CMakeFiles/FreeCADGui.dir/DlgSettingsMacroImp.cpp.o
/home/wile/Source/FreeCAD/src/Gui/DlgGeneralImp.cpp: In member function ‘void Gui::Dialog::DlgGeneralImp::emitReloadWorkbenchNeeded(bool)’:
/home/wile/Source/FreeCAD/src/Gui/DlgGeneralImp.cpp:251:52: warning: unused parameter ‘val’ [-Wunused-parameter]
  251 | void DlgGeneralImp::emitReloadWorkbenchNeeded(bool val) {

**Comment by PaddleStroke** [comment]
Thanks for looking into it! I think you grabbed my wrong branch.
Wb_cornerwidget had a problem, I closed the PR of that branch and opened a
new branch/PR see the branch of this PR, it's corner_widget_test I seem to
recall

**Comment by PaddleStroke** [comment]
Edit: The correct branch is [wb_corner_test](https://github.com/PaddleStroke/FreeCAD/tree/wb_corner_test)

**Comment by 0penBrain** [comment]
@WandererFan Why are you not just checking out the PR itself rather than the contributor fork branch ?

**Comment by PaddleStroke** [comment]
> My favorite choice would be to have an button before the file menu with the icon of the active wb, opening a menu to access the list and change wb...

This is possible. I actually made it that way at some point. And it would work with KDE Global menu. My only concern was that it was not much proiminent like this. It felt a bit hidden. I wanted the Icon + workbench name + Color of menu background change, to make it proiminent. But that not possible with QT.

**Comment by WandererFan** [comment]
> @WandererFan Why are you not just checking out the PR itself rather than the contributor fork branch ?

Lack of expertise.  How does one do this?

**Comment by 0penBrain** [comment]
> Lack of expertise. How does one do this?

Guessing 'origin' is your own fork and 'upstream' the main FC repo -- I'll take this PR (7679) as example --:
```
git fetch upstream pull/7679/head:PR7679 # 'PR7679' at the end is a free name, reused then
git checkout PR7679
```
Then later if you need to update, just:
```
git pull --ff-only upstream pull/7679/head
```

**Comment by PaddleStroke** [comment]
> [Merge branch 'master' into wb_corner_test](https://github.com/FreeCAD/FreeCAD/pull/7679/commits/711eb7683ff9f0cef7b8587ce138e5fcbdb9090a)
I think there might have been a mishap here (unless I'm missing something which is totally possible). Do you want me to rebase on latest master or did you intend to merge the PR in master?

**Comment by 0penBrain** [comment]
@WandererFan Would you have inadvertently 'push' something after fetching the PR ?

**Comment by WandererFan** [comment]
I pressed the github button for "This branch is out-of-date with the base branch
Merge the latest changes from master into this branch."

File: src/Gui/Action.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (wwmayer) [reply src/Gui/Action.cpp]: In new code the old-style connect() shouldn't be used any more. Use the new-style version instead. Because onActivated is a member function of WorkbenchComboBox the argument of the lambda expression must be adjusted.

**Comment by silopolis** [comment]
Le lun. 7 nov. 2022 à 11:47, PaddleStroke ***@***.***> a
écrit :

> My favorite choice would be to have an button before the file menu with
> the icon of the active wb, opening a menu to access the list and change
> wb...
>
> This is possible. I actually made it that way at some point. And it would
> work with KDE Global menu.
>

Even better!
Maybe as an option 🙏 😋

**Comment by luzpaz** [comment]
> I pressed the github button for "This branch is out-of-date with the base branch
> Merge the latest changes from master into this branch."

Hey @WandererFan I recommend not using github UI for this at all. It's does unexpected changes in my experience.

File: src/Gui/DlgGeneralImp.cpp [hunk 3]
```diff
@@ -215,6 +215,25 @@ void DlgGeneralImp::saveSettings()
     hGrp->GetGroup("TreeView")->SetBool("Enabled",treeView);
     hGrp->GetGroup("PropertyView")->SetBool("Enabled",propertyView);
 
+    //save workbench selector position
+    hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/MainWindow");
+    bool WSToolbar = false, WSLeftCorner = false, WSRightCorner = false;
```
Comment (wwmayer) [inline src/Gui/DlgGeneralImp.cpp:220]: This whole logic is weird because the 3 boolean are actually mutual exclusive but the way how it's done here they can be combined because each parameter can be set to true or false independently (directly in the user.cfg).
This results into the weird behaviour that if the keys WSLeftCorner and WSRightCorner are set to true two switcher appear, one on the left side and that in the toolbar (not that on the right side)

The three parameters should be replaced by one string parameter called WSPosition and its values are "WSToolbar", "WSLeftCorner" and "WSRightCorner". The default or in case of an invalid value "WSToolbar" is used.

Comment (PaddleStroke) [reply src/Gui/DlgGeneralImp.cpp:220]: Thanks.
I understand.
So this + change the connect to the new style. Is that all?
Do you want to push directly or should I make a PR for those changes?

**Comment by PaddleStroke** [comment]
@WandererFan thanks for looking into it :)

File: src/Gui/DlgGeneralImp.cpp (hunk 3 above)
Comment (0penBrain) [reply src/Gui/DlgGeneralImp.cpp:220]: @wwmayer how about using a `PrefComboBox`? This class today stores the index of the selected item, which is a bit boring because then in the code it's like dealing with magic numbers. But it exists already and at some point we can maybe improve it to store a user-defined value instead of index.

Comment (wwmayer) [reply src/Gui/DlgGeneralImp.cpp:220]: @0penBrain the difficulty with PrefComboBox is that the string of an item cannot be used because this usually is translated text and changing the language will break the reading of the parameter. The alternative is to use a combo box' user data but in most cases a programmer doesn't set it.

I think the best we can do is defining an enum and a map or array of QString with the translated text. The index of the combo box is cast to the enum.

@PaddleStroke 
> So this + change the connect to the new style. Is that all? Do you want to push directly or should I make a PR for those changes?

Yes, that's it. And yes I will have a look...

Comment (wwmayer) [reply src/Gui/DlgGeneralImp.cpp:220]: See #7744

**Comment by 0penBrain** [comment]
Wiki pages about Preferences and Release notes to be updated.

**Comment by PaddleStroke** [comment]
done: 
https://wiki.freecadweb.org/Release_notes_1.0
https://wiki.freecadweb.org/Preferences_Editor
